src/core/Akka.Streams/Dsl/Restart.cs

```diff
@@ -0,0 +1,490 @@
+                    }
+                }));
+
+            SetHandler(Out, 
+                onPull: () => sinkIn.Pull(),
+                onDownstreamFinish: () =>
+                {
+                    _finishing = true;
+                    sinkIn.Cancel();
+                });
+            
+            return sinkIn;
+        }
+
+        protected SubSourceOutlet<T> CreateSubOutlet<T>(Inlet<T> inlet)
+        {
+            var sourceOut = new SubSourceOutlet<T>(this, $"RestartWithBackoff{_name}.subOut");
+            sourceOut.SetHandler(new LambdaOutHandler(
+                onPull: () =>
+                {
+                    if (IsAvailable(In))
+                    {
+                        sourceOut.Push(Grab<T>(In));
+                    }
+                    else
+                    {
+                        if (!HasBeenPulled(In))
+                            Pull(In);
+                    }
+                },
+                onDownstreamFinish: () =>
+                {
+                    if (_finishing)
+                    {
+                        Cancel(In);
+                    }
+                    else
+                    {
+                        Log.Debug("Graph in finished");
+                        OnCompleteOrFailure();
+                    }
+                }
+            ));
+
+            SetHandler(In, 
+                onPush: () =>
+                {
+                    if (sourceOut.IsAvailable)
+                        sourceOut.Push(Grab<T>(In));
+                },
+                onUpstreamFinish: () =>
+                {
+                    _finishing = true;
+                    sourceOut.Complete();
+                },
+                onUpstreamFailure: ex =>
+                {
+                    _finishing = true;
+                    sourceOut.Fail(ex);
+                });
+
+            return sourceOut;
+        }
+
+        internal void OnCompleteOrFailure()
+        {
+            // Check if the last start attempt was more than the minimum backoff
+            if (_resetDeadline.IsOverdue)
+            {
+                Log.Debug($"Last restart attempt was more than {_minBackoff} ago, resetting restart count");
+                _restartCount = 0;
+            }
+
+            var restartDelay = BackoffSupervisor.CalculateDelay(_restartCount, _minBackoff, _maxBackoff, _randomFactor);
+            Log.Debug($"Restarting graph in {restartDelay}");
+            ScheduleOnce("RestartTimer", restartDelay);
+            _restartCount += 1;
+            // And while we wait, we go into backoff mode
+            Backoff();
+        }
+
+        protected internal override void OnTimer(object timerKey)
+        {
+            StartGraph();
+            _resetDeadline = _minBackoff.FromNow();
+        }
+
+        // When the stage starts, start the source
+        public override void PreStart() => StartGraph();
+    }
+
+    internal sealed class Deadline
```

Comment:
We have a similar class in Akka.Remote, but it accepts `DateTime`, not `TimeSpan`

Reply:
@Aaronontheweb What to do with this class?